File: apps/python/lib/chapter08/__main__.py
from pathlib import Path

import pandas as pd

from lib.chapter02.iris_preprocessing import split_train_test
from lib.chapter08.survived_classifier import (
    FEATURES,
    MODEL_FILE,
    build_pipeline,
    evaluate,
    load_model,
    load_survived,
    save_model,
    split_features_and_target,
)
from lib.dataset import data_dir

TEST_SIZE = 0.2
SEED = 0
MAX_DEPTH = 5

NEW_PASSENGERS = pd.DataFrame(
    [
        (1, "female", None, 0, 0, 50.0, "C"),
        (3, "male", None, 0, 0, 8.0, "S"),
    ],
    columns=FEATURES,
)


def main(model_file: Path = MODEL_FILE) -> None:
    df = load_survived(data_dir() / "Survived.csv")
    x, t = split_features_and_target(df)
    split = split_train_test(x, t, test_size=TEST_SIZE, seed=SEED)
    counts = t.value_counts()
    print(f"データ件数: {len(df)}（生存 {counts[1]}, 死亡 {counts[0]}）")
    print(f"訓練データ: {len(split.x_train)} 件, テストデータ: {len(split.x_test)} 件")

    pipelines = {}
    for class_weight in [None, "balanced"]:
        pipeline = build_pipeline(max_depth=MAX_DEPTH, class_weight=class_weight)
        pipelines[class_weight] = pipeline.fit(split.x_train, split.t_train)
        result = evaluate(pipeline, split)
        print(
            f"class_weight={class_weight}: "
            f"訓練 {result.train_accuracy:.3f}, テスト {result.test_accuracy:.3f}, "
            f"生存者 {result.survivors} 人中 {result.found_survivors} 人を発見"
        )

    save_model(pipelines["balanced"], model_file)
    predictions = load_model(model_file).predict(NEW_PASSENGERS)
    print(f"保存したモデル: {model_file.name}")
    print(f"架空の乗客の予測: {predictions.tolist()}")


if __name__ == "__main__":
    main()

File: apps/kotlin/src/main/kotlin/chapter08/Main.kt
package chapter08

import chapter02.splitTrainTest
import dataset.dataDir
import org.jetbrains.kotlinx.dataframe.AnyFrame
import org.jetbrains.kotlinx.dataframe.api.dataFrameOf
import java.io.File
import java.util.Locale

private const val TEST_SIZE = 0.2
private const val SEED = 0
private const val MAX_DEPTH = 5

/** 学習済みのパイプラインの保存先（apps/kotlin/model/ は .gitignore の対象） */
val MODEL_FILE = File("model/survived.ser")

private val NEW_PASSENGERS: AnyFrame =
    dataFrameOf(
        "Pclass" to listOf(1, 3),
        "Sex" to listOf("female", "male"),
        "Age" to listOf(null, null),
        "SibSp" to listOf(0, 0),
        "Parch" to listOf(0, 0),
        "Fare" to listOf(50.0, 8.0),
        "Embarked" to listOf("C", "S"),
    )

private fun format(value: Double): String = "%.3f".format(Locale.ROOT, value)

fun main() = main(MODEL_FILE)

fun main(modelFile: File) {
    val df = loadSurvived(File(dataDir(), "Survived.csv"))
    val (x, t) = splitFeaturesAndTarget(df)
    val split = splitTrainTest(x, t, testSize = TEST_SIZE, seed = SEED)
    val counts = t.groupingBy { it }.eachCount()
    println("データ件数: ${df.rowsCount()}（生存 ${counts[1]}, 死亡 ${counts[0]}）")
    println("訓練データ: ${split.xTrain.rowsCount()} 件, テストデータ: ${split.xTest.rowsCount()} 件")

    val pipelines =
        ClassWeight.entries.associateWith { classWeight ->
            buildPipeline(maxDepth = MAX_DEPTH, classWeight = classWeight).fit(split.xTrain, split.tTrain)
        }
    for ((classWeight, pipeline) in pipelines) {
        val result = evaluate(pipeline, split)
        println(
            "classWeight=$classWeight: 訓練 ${format(result.trainAccuracy)}, テスト ${format(result.testAccuracy)}, " +
                "生存者 ${result.survivors} 人中 ${result.foundSurvivors} 人を発見",
        )
    }

    saveModel(pipelines.getValue(ClassWeight.BALANCED), modelFile)
    val predictions = loadModel(modelFile).predict(NEW_PASSENGERS)
    println("保存したモデル: ${modelFile.name}")
    println("架空の乗客の予測: $predictions")
}
